pkg/cmd/kind/test/internal/testcmd/testcmd.go

```diff
@@ -0,0 +1,363 @@
+		_ = exec.Command("create_bazel_cache_rcs.sh").Run()
+	}
+
+	// first build the node image
+	ctx, err := node.NewBuildContext(
+		node.WithMode(buildType),
+		node.WithImage(flags.ImageName),
+		node.WithLogger(logger),
+	)
+	if err != nil {
+		return errors.Wrap(err, "error creating build context")
+	}
+	if err := ctx.Build(); err != nil {
+		return errors.Wrap(err, "error building node image")
+	}
+
+	// build test binaries
+	switch buildType {
+	case "bazel":
+		if err := exec.InheritOutput(
+			exec.Command("bazel", "build", "//test/e2e:e2e.test", "//vendor/github.com/onsi/ginkgo/ginkgo"),
+		).Run(); err != nil {
+			return err
+		}
+	case "docker":
+		if err := exec.InheritOutput(
+			exec.Command("make", "all", `WHAT=cmd/kubectl test/e2e/e2e.test vendor/github.com/onsi/ginkgo/ginkgo`),
+		).Run(); err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("build mode %q is not implemented yet!", buildType)
+	}
+
+	// In Kubernetes's CI: attempt to release some memory after building
+	if len(os.Getenv("KUBETEST_IN_DOCKER")) > 0 {
+		_ = exec.Command("sync").Run()
+		_ = exec.Command("/proc/sys/vm/drop_caches").SetStdin(strings.NewReader("1")).Run()
+	}
+
+	return nil
+}
+
+func checkAlreadyExists(provider *cluster.Provider, name string) error {
+	n, err := provider.ListNodes(name)
+	if err != nil {
+		return err
+	}
+	if len(n) != 0 {
+		return fmt.Errorf("node(s) already exist for a cluster with the name %q", name)
+	}
+	return nil
+}
+
+func testGinkgo(logger log.Logger, provider *cluster.Provider, flags *flagpole, details Details) error {
+	// construct the cluster config
+	config := &v1alpha4.Cluster{
+		Networking: v1alpha4.Networking{
+			IPFamily: v1alpha4.ClusterIPFamily(flags.IPFamily),
+		},
+		Nodes: []v1alpha4.Node{
+			{
+				Role:  v1alpha4.ControlPlaneRole,
+				Image: flags.ImageName,
+			},
+		},
+	}
+	// add e2e workers
+	numNodesForE2E := 2
+	for i := 0; i < numNodesForE2E; i++ {
+		config.Nodes = append(config.Nodes, v1alpha4.Node{
+			Role:  v1alpha4.WorkerRole,
+			Image: flags.ImageName,
+		})
+	}
+
+	// create the cluster
+	logger.V(0).Infof("Creating cluster %q ...\n", flags.Name)
+	if err := provider.Create(
+		flags.Name,
+		cluster.CreateWithV1Alpha4Config(config),
+		cluster.CreateWithRetain(true),
+		cluster.CreateWithWaitForReady(time.Minute),
+		cluster.CreateWithKubeconfigPath(flags.kubeconfigPath()),
+		cluster.CreateWithDisplayUsage(false),
+		cluster.CreateWithDisplaySalutation(false),
+	); err != nil {
+		if errs := errors.Errors(err); errs != nil {
+			for _, problem := range errs {
+				logger.Errorf("%v", problem)
+			}
+			return errors.New("aborting due to invalid configuration")
+		}
+		return errors.Wrap(err, "failed to create cluster")
+	}
+
+	// In Kubernetes CI we don't have real IPv6 connectivity so we need to
+	// employ some work arounds when testing IPv6 clusters
+	// We do this all environments for consistency.
+	if config.Networking.IPFamily == v1alpha4.IPv6Family {
```

Comment:
`config.Networking.IPFamily == v1alpha4.IPv6Family || config.Networking.IPFamily == v1alpha4.DualStackFamily`

Reply:
we can't realy TODO constants that don't exist yet, dual stack my not be a family, it will probably be many families?
dual stack as an API is a TODO.

also, there is validation long before this code rune that errors if not one of ipv4 / ipv6 specifically to handle the fact that this command is not aware of any other families currently

Reply:
👍